dissent-oracle/contracts/DissentOracle.sol

```diff
@@ -0,0 +1,79 @@
+pragma solidity ^0.4.24;
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+import "@aragon/os/contracts/acl/IACLOracle.sol";
+import "@aragon/os/contracts/lib/math/SafeMath64.sol";
+import "../../dandelion-voting/contracts/DandelionVoting.sol";
+
+contract DissentOracle is AragonApp, IACLOracle {
+
+    using SafeMath64 for uint64;
+
+    bytes32 public constant SET_DANDELION_VOTING_ROLE = keccak256("SET_DANDELION_VOTING_ROLE");
+    bytes32 public constant SET_DISSENT_WINDOW_ROLE = keccak256("SET_DISSENT_WINDOW_ROLE");
+
+    string private constant ERROR_NOT_CONTRACT = "DISSENT_ORACLE_NOT_CONTRACT";
+
+    DandelionVoting public dandelionVoting;
+    uint64 public dissentWindowBlocks;
+
+    event DandelionVotingSet(address dandelionVoting);
+    event DissentWindowBlocksSet(uint64 dissentWindowBlocks);
+
+    /**
+    * @notice Initialize the DissentOracle
+    * @param _dandelionVoting Dissent voting aragon app address
+    * @param _dissentWindowBlocks Blocks from previous yea vote the oracle will approve a function call for
+    */
+    function initialize(address _dandelionVoting, uint64 _dissentWindowBlocks) external onlyInit {
+        require(isContract(_dandelionVoting), ERROR_NOT_CONTRACT);
+
+        dandelionVoting = DandelionVoting(_dandelionVoting);
+        dissentWindowBlocks = _dissentWindowBlocks;
+
+        initialized();
+    }
+
+    /**
+    * @notice Update the dissent voting app to `_dandelionVoting`
+    * @param _dandelionVoting New dissent window
+    */
+    function setDandelionVoting(address _dandelionVoting) external auth(SET_DANDELION_VOTING_ROLE) {
+        require(isContract(_dandelionVoting), ERROR_NOT_CONTRACT);
+
+        dandelionVoting = DandelionVoting(_dandelionVoting);
+        emit DandelionVotingSet(_dandelionVoting);
+    }
+
+    /**
+    * @notice Update the dissent window to `_dissentWindowBlocks`
+    * @param _dissentWindowBlocks New dissent window
+    */
+    function setDissentWindow(uint64 _dissentWindowBlocks) external auth(SET_DISSENT_WINDOW_ROLE) {
+        dissentWindowBlocks = _dissentWindowBlocks;
+        emit DissentWindowBlocksSet(_dissentWindowBlocks);
+    }
+
+    /**
+    * @notice ACLOracle
+    * @dev IACLOracle interface conformance. If the ACLOracle permissioned function uses a forwarder and needs to
+    *      pass parameters, it should be used with the modifier 'authP(SOME_ACL_ROLE, arr(voter, dissentWindowBlocks))'
+    */
+    function canPerform(address _who, address _where, bytes32 _what, uint256[] _how) external view returns (bool) {
+
+        address voter = _who;
+        uint64 dissentWindowBlocksLocal = dissentWindowBlocks;
+
+        if (_how.length > 0) {
+            voter = address(_how[0]);
+            dissentWindowBlocksLocal = uint64(_how[1]);
+        }
```

Comment:
I've removed the ability to pass an address through the params. I decided if we want to use the Dissent Oracle with a forwarder we should update it then and the params should be the `msgHash` and `signature` so the Oracle can verify the address itself. Less security risks that way.

I've also debated getting rid of the ability to pass the `dissentWindowBlocks` in the params. If it is used it might become a necessary param if wanting to add forwarder functionality with the extra params `msgHash` and `signature`. Which wouldn't be ideal. However, their might be use cases for specifying different dissent windows so I'll leave it for now.

Reply:
Yeah, i agree passing multiple values as params can become cumbersome in some cases.

If in the future we'll want to use redemptions back through a forwarder, the approach that you mentioned is the way to go. In that case i think we can check the params separately so it's not a requirement to also have the dissentWindowBlocks in the function modifier `authP(ROLE, arr())` but the general problem here is that we cannot have the `nth + 1` param without having the `nth`.

Reply:
> the general problem here is that we cannot have the `nth + 1` param without having the `nth`

Yup, well explained. Getting rid for now anyway.